Promscale - Add Overview & Getting Started (#478)
timescale/docs: promscale/installation/docker.md

# Test Promscale using a pre-built container
If you want to try out Promscale in a test environment before getting started,
you can use a pre-built Docker container.

<highlight type="important">
Using Promscale from a Docker container is for local testing purposes only. Do

> not use this installation method in a production environment.

Reviewer: I'm missing a "Why/why not?" here. Without that it sounds like I should be afraid, but not of what.

Reviewer: I want to add: I assume that the issue here is data retention i.e. customers may shoot themselves in the foot and lose their data. If that's the case, say so. If it's something more than that then we should also explicitly mention what it is.

Author: @VineethReddy02 ?

Reviewer 2: This statement is part of our GitHub docker docs. So copy pasting it added it here. 

IMO the docker setup we suggested is two modes setting up individual containers or using the docker-compose file. Usually the day 2 operations such as scaling, upgrading are troublesome and involve multiple failure points. So assuming that's the reason for not recommending prod use cases.

Reviewer: I tried to figure out where it was copy-pasted from, but I couldn't. The closest thing I could find was the warning above this section (scroll up a little): https://docs.timescale.com/timescaledb/latest/tutorials/promscale/promscale-install/#install-timescaledb.

That message and this one are not consistent.

On that page it says (reading between the lines, emphasis mine): "these _instructions_ are not necessarily suitable for a production setup", whereas here we are saying (reading between the lines, and rephrasing): "running Promscale in a Docker container is not suitable for production".

Reviewer 3: I rewrote the page because I noticed the same thing. It should be good now. Let us know if you think it's still unclear.

</highlight>

## Install Promscale with Docker
Before you begin, you must have Docker installed on your local system. For
packages and instructions, see the
[Docker installation documentation][docker-install].

<procedure>

### Installing Promscale with Docker
1.  Use Docker to create a network for Promscale and TimescaleDB:
    ```bash
    docker network create --driver bridge promscale-timescaledb
    ```

> 1.  Install TimescaleDB in a Docker container, and make sure the service is

Reviewer: This step and the next one are confused. The code in the next step fits to this step, and the next step is missing its code.

Author: @VineethReddy02 ?

Reviewer 2: updated.

    running. For instructions, see

>     [the TimescaleDB Docker installation section][tsdb-docker]. Promscale stores

Reviewer: broken link

Reviewer: This link is still broken

    all metrics data scraped from Prometheus targets in this TimescaleDB service.
1.  Run the Promscale Docker container. This example creates a TimescaleDB

>     instance called `timescaledb`, on a network named `promscale-timescale`. The

Reviewer: ```suggestion
    instance called `timescaledb`, on a network named `promscale-timescaledb`. The
```

        --network promscale-timescaledb \
        -e POSTGRES_PASSWORD=<password> -d -p 5432:5432 \
        timescaledev/timescaledb-ha:pg12-latest
    ```

</procedure>

## Collect metrics with node_exporter
When you have Promscale running in a container, you can begin collecting metrics. The `node_exporter` tool is a service that exposes hardware and operating systems metrics to Prometheus. For more information about `node_exporter`, see the [node_exporter developer documentation][gh-node-exporter].

<procedure>

### Collecting metrics with node_exporter
1.  Run the `node_exporter` Docker container. This example creates
    a `node_exporter` instance called `node_exporter`, port forwarding output to
    port `9100`, and it runs on the `promscale-timescaledb` network you created
    earlier:
    ```bash
    docker run --name node_exporter -d -p 9100:9100 \
    --network promscale-timescaledb \
    quay.io/prometheus/node-exporter
    ```
1.  Verify that system metrics are being exported, by opening
    `http://localhost:9100/metrics` in your browser. This is the location that
    Prometheus scrapes for metrics data.

</procedure>

## Install Prometheus

> When you are succecssdfully receiving metrics from your Promscale instalaltion, you can install Prometheus and configure it to garther the metrics data.

Reviewer: Aside from the typos in this sentence, the content is confusing: "when you are successfully receiving metrics...". At this point in the installation process, how should I know if that is the case?

I would suggest replacing this sentence with:

Now that you are successfully generating metrics in `node_exporter`, you can install Prometheus, configure it to collect metrics data from the `node_exporter`, and forward those metrics to Promscale.

Reviewer 3: This content was removed so the problem doesn't exist anymore

<procedure>

### Installing Prometheus

> 1.  Create a Prometheus configuration file called `prometheus.yml`. You can use

Reviewer: ```suggestion
1.  Create a Prometheus configuration file `prometheus.yml`. You can use
```

Reviewer 3: This content was removed so the problem doesn't exist anymore

>     this example as a starting point. In this example, Prometheus uses Promscale

Reviewer: What does "use this example as a starting point" mean? We should provide a complete and working prometheus config here and not have any questions about completeness.

Reviewer 2: Usually, the Prometheus config is custom tailor-made for one's requirements. For example the `scrape_interval`, `targets`, `recording_rules`, `alerting_rules1`, etc... 

So we do not want to guarantee or recommend using the example/sample config shared by us. :) 
The main intention of adding the Prometheus section is to help a fraction of users who are new to Prometheus to get started with the stack.

Reviewer 3: This content was removed so we no longer have this issue

    Additionally, it sets `node_exporter` as the target, and scrapes every
    10&nbsp;seconds:
    ```yaml
    global:
     scrape_interval:     10s
     evaluation_interval: 10s
    scrape_configs:
     - job_name: prometheus
       static_configs:
         - targets: ['localhost:9090']
     - job_name: node-exporter
       static_configs:
         - targets: ['node_exporter:9100']
    remote_write:
      - url: "http://promscale:9201/write"
    remote_read:
      - url: "http://promscale:9201/read"
        read_recent: true
    ```
1.  Start the Prometheus Docker container, using the configuration file you
    created:
    ```bash
    docker run \
        --network promscale-timescaledb \
        -p 9090:9090 \
        -v ${PWD}/prometheus.yml:/etc/prometheus/prometheus.yml \
        prom/prometheus
    ```

</procedure>

## Install Promscale with a Docker compose file
You can use a Docker compose file to create containers for TimescaleDB, Promscale, and the `node_exporter`, using a single command. To help you get started, we have created a sample Docker compose file, available from the [Promscale GitHub repository][promscale-docker-compose].

> Before you begin, make sure that you have the `prometheus.yml` configuration file in the same directory as `docker-compose.yml`.

Reviewer: This refers to the config file in the previous section. It's possible that readers will jump straight here and not see that, and wonder what is meant.

# Test Promscale using a pre-built container
If you want to try out Promscale in a test environment before getting started,
you can use a pre-built Docker container.

<highlight type="important">

> Running Promscale in a Docker container is not suitable for production use-cases. 

Author: ```suggestion
The instructions given in this section are designed for you to get started with Promscale quickly. They are especially suited for creating a proof of concept. We do not recommend that you run production workloads with Promscale in a Docker container. 
```

</highlight>

## Install Promscale with Docker
Before you begin, you must have Docker installed on your local system. For
packages and instructions, see the
[Docker installation documentation][docker-install].

<procedure>

### Installing Promscale with Docker
1.  Use Docker to create a network for Promscale and TimescaleDB:
    ```bash
    docker network create --driver bridge promscale
    ```
1.  Install TimescaleDB in a Docker container on a network named `promscale`. It also port forwards to port `5432` on your local system:
    ```bash
    docker run --name timescaledb -e POSTGRES_PASSWORD=<password> -d -p 5432:5432 --network promscale timescaledev/promscale-extension:latest-ts2-pg13 postgres -csynchronous_commit=off
    ```
1.  Run the remote-storage connector Docker container on a network named `promscale`. It also port forwards to port `9201` on your local system:
    ```bash

>     docker run --name promscale -d -p 9201:9201 --network promscale timescale/promscale:latest -db-password=<password> -db-port=5432 -db-name=postgres -db-host=timescaledb -db-ssl-mode=allow

Author: ```suggestion
    docker run --name promscale -d -p 9201:9201 \
    --network promscale timescale/promscale:latest \
    -db-password=<password> \
    -db-port=5432 \
    -db-name=postgres \
    -db-host=timescaledb \
    -db-ssl-mode=allow
```

        prom/prometheus
    ```

</procedure>

## Alternatively you can install all the above components with a Docker compose file

You can use a Docker compose file to create containers for TimescaleDB,
Promscale, and the NodeExporter, using a single command. To help you get
started, we have created a sample Docker compose file, available from the
[Promscale GitHub repository][promscale-docker-compose].

Before you begin, make sure that you have your `prometheus.yml` configuration
file in the same directory as `docker-compose.yml`. For more information
about creating a `prometheus.yml` configuration file, see the
[install Prometheus][install-prometheus] step in this section.

<procedure>

### Installing Promscale with a Docker compose file
1.  Open the `docker-compose.yml` configuration file, and set the password for
    the `postgres` user.
1.  In the directory that conatins the `docker-compose.yml` file, run the
    `docker-compose` command:
    ```bash
    docker-compose up
    ```

</procedure>

[promscale-docker-compose]: https://github.com/timescale/promscale/blob/master/docker-compose/docker-compose.yaml
[docker-install]: https://docs.docker.com/get-docker/

> [tsdb-docker]: timescaledb/:currentVersion:/how-to-guides/install-timescaledb/self-hosted/docker/installation-docker/

Author: ```suggestion
```

you can use a pre-built Docker container.

<highlight type="important">

> Running Promscale in a Docker container is not suitable for production use-cases. 

Reviewer 3: Check the page now: https://docs.timescale.com/promscale/latest/installation/docker/

It first points to the images, then tells people to refer to the instructions of their own container platform or our Kubernetes instructions. And finally, we say
"Running Promscale directly using docker run is not recommended for production environments. This can be useful for testing purposes and is just provided as an example."

### Installing Promscale with Docker
1.  Use Docker to create a network for Promscale and TimescaleDB:
    ```bash
    docker network create --driver bridge promscale
    ```
1.  Install TimescaleDB in a Docker container on a network named `promscale`. It also port forwards to port `5432` on your local system:
    ```bash

>     docker run --name timescaledb -e POSTGRES_PASSWORD=<password> -d -p 5432:5432 --network promscale timescaledev/promscale-extension:latest-ts2-pg13 postgres -csynchronous_commit=off

Author: Using <REPLACEABLE_TEXT> is standard in tech docs. The idea being that when people copy and paste, the command will fail until they have replaced it. I think allowing people to copy paste without changing things that need changing is inadvisable.

>     docker run --name timescaledb -e POSTGRES_PASSWORD=<password> -d -p 5432:5432 --network promscale timescaledev/promscale-extension:latest-ts2-pg13 postgres -csynchronous_commit=off

Reviewer: @VineethReddy02 Can you comment on where `synchronous_commit=off` comes from? Is that a recommended setting?

Could we also line-break this command so that it's completely visible on the docs page?

>     docker run --name timescaledb -e POSTGRES_PASSWORD=<password> -d -p 5432:5432 --network promscale timescaledev/promscale-extension:latest-ts2-pg13 postgres -csynchronous_commit=off

Reviewer 3: I agree with both of you :) Some docs have a button to auto-generate a password so the user can just copy and paste so the password is random and secure and they know it was auto-generated because they had to take action.

>     docker run --name timescaledb -e POSTGRES_PASSWORD=<password> -d -p 5432:5432 --network promscale timescaledev/promscale-extension:latest-ts2-pg13 postgres -csynchronous_commit=off

Reviewer 3: So I'd say for now we leave as is.

>     docker run --name timescaledb -e POSTGRES_PASSWORD=<password> -d -p 5432:5432 --network promscale timescaledev/promscale-extension:latest-ts2-pg13 postgres -csynchronous_commit=off

Reviewer 3: AFAIK synchronous_commit=off is a recommended setting and we use it in Timescale Cloud for example.

>     docker run --name timescaledb -e POSTGRES_PASSWORD=<password> -d -p 5432:5432 --network promscale timescaledev/promscale-extension:latest-ts2-pg13 postgres -csynchronous_commit=off

Reviewer 3: We've broken up one line (the Promscale one) but not the other. PR added: https://github.com/timescale/docs/pull/684

    ```
1.  Run the remote-storage connector Docker container on a network named `promscale`. It also port forwards to port `9201` on your local system:
    ```bash

>     docker run --name promscale -d -p 9201:9201 --network promscale timescale/promscale:latest -db-password=<password> -db-port=5432 -db-name=postgres -db-host=timescaledb -db-ssl-mode=allow

Reviewer 3: This was done

    created:
    ```bash
    docker run \
        --network promscale \
        -p 9090:9090 \

>         -v ${PWD}/prometheus.yml:/etc/prometheus/prometheus.yml \

Reviewer 3: Prometheus content was removed so no longer an issue

        prom/prometheus
    ```

</procedure>

## Alternatively you can install all the above components with a Docker compose file

You can use a Docker compose file to create containers for TimescaleDB,

> Promscale, and the NodeExporter, using a single command. To help you get

Reviewer 3: Prometheus content was removed so no longer an issue

</procedure>

## Install Promscale with a Docker compose file
You can use a Docker compose file to create containers for TimescaleDB,
Promscale, and the `node_exporter`, using a single command. To help you get
started, we have created a sample Docker compose file, available from the
[Promscale GitHub repository][promscale-docker-compose].

Before you begin, make sure that you have your `prometheus.yml` configuration

> file is in the same directory as `docker-compose.yml`. For more information

Reviewer: ```suggestion
file in the same directory as `docker-compose.yml`. For more information
```

Reviewer 3: This content was removed so the problem doesn't exist anymore

about creating a `prometheus.yml` configuration file, see the

> [install Prometheus][install-prometheus] step in this section.

Reviewer: broken link

[promscale-docker-compose]: https://github.com/timescale/promscale/blob/master/docker-compose/docker-compose.yaml
[docker-install]: https://docs.docker.com/get-docker/

> [tsdb-docker]: timescaledb/:currentVersion:/how-to-guides/install-timescaledb/self-hosted/docker/installation-docker/

Reviewer 3: This was fixed
